owca/allocations.py

```diff
@@ -0,0 +1,197 @@
+    extend them with necessary business logic for:
+    - calculating changeset for comparing dict like containers
+    - recursive validation of all values of dict
+    - collection of metrics of all values
+    - and recursive perform allocations
+    """
+
+    def calculate_changeset(self, current: 'AllocationsDict') \
+            -> Tuple['AllocationsDict', Optional['AllocationsDict']]:
+        assert isinstance(current, AllocationsDict)
+
+        # Create an shallow copy of current object that will represent 'sum'.
+        target = AllocationsDict(current)
+        # Empty object to represnt nessesary changes to apply.
+        changeset = AllocationsDict({})
+
+        for key, new_value in self.items():
+
+            assert isinstance(new_value, AllocationValue)
+
+            current_value = current.get(key)
+
+            if current_value is None:
+                # There is no current value, new is used as both target and changeset.
+                target[key] = new_value
+                changeset[key] = new_value
+            else:
+                # Both exists - recurse into values.
+                assert isinstance(current_value, AllocationValue)
+                target_value, value_changeset = new_value.calculate_changeset(current_value)
+                assert isinstance(target_value, AllocationValue)
+                assert isinstance(value_changeset, (type(None), AllocationValue))
+                target[key] = target_value
+                if value_changeset is not None:
+                    changeset[key] = value_changeset
+
+        # If there is no fields in changeset dict return None
+        # to indicate no changes are required at all.
+        if not changeset:
+            changeset = None
+
+        return target, changeset
+
+    def generate_metrics(self) -> List[Metric]:
+        metrics = []
+        for value in self.values():
+            metrics.extend(value.generate_metrics())
+        return metrics
+
+    def perform_allocations(self):
+        for value in self.values():
+            value.perform_allocations()
+
+    def validate(self):
+        for value in self.values():
+            value.validate()
+
+
+class LabelsUpdater:
+    """Helper object to update metrics."""
+
+    def __init__(self, common_labels):
+        self.common_labels = common_labels
+
+    def update_labels(self, metrics):
+        """Update labels values inplace."""
+        for metric in metrics:
+            metric.labels.update(**self.common_labels)
+
+
+class BoxedNumeric(AllocationValue):
+    """ AllocationValue for numeric like values (floats and ints).
+    Wrapper for floats and integers.
+    If min_value is None then it becomes negative infinity (default is 0).
+    If max_value is None then it becomes infinity (default is None(infinity).
+    """
+    # Defines default how sensitive in terms of
+    # float precision are changes are detected.
+    FLOAT_VALUES_CHANGE_DETECTION = 0.05
```

Comment:
```suggestion
    VALUE_CHANGE_SENSITIVITY = 0.05
```
(you will use it to compare integers too).